crates/wasi-common/src/fdentry.rs

```diff
@@ -2,40 +2,139 @@ use crate::sys::dev_null;
+            Handle::VirtualFile(virt) => virt.try_clone().map(|f| Descriptor::VirtualFile(f)),
+        }
+    }
+}
+
+impl<'descriptor> fmt::Debug for Handle<'descriptor> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Handle::OsHandle(file) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = file;
+                write!(f, "{:?}", file)
+            }
+            Handle::Stream(stream) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = stream;
+                write!(f, "{:?}", file)
+            }
+            Handle::VirtualFile(_) => write!(f, "VirtualFile"),
+        }
+    }
+}
 
-#[derive(Debug)]
 pub(crate) enum Descriptor {
     OsHandle(OsHandle),
+    VirtualFile(Box<dyn VirtualFile>),
     Stdin,
     Stdout,
     Stderr,
 }
 
+impl fmt::Debug for Descriptor {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Descriptor::OsHandle(handle) => write!(f, "{:?}", handle),
+            Descriptor::VirtualFile(_) => write!(f, "VirtualFile"),
+            Descriptor::Stdin => write!(f, "Stdin"),
+            Descriptor::Stdout => write!(f, "Stdout"),
+            Descriptor::Stderr => write!(f, "Stderr"),
+        }
+    }
+}
+
 impl Descriptor {
-    /// Return a reference to the `OsHandle` treating it as an actual file/dir, and
-    /// allowing operations which require an actual file and not just a stream or
-    /// socket file descriptor.
-    pub(crate) fn as_file(&self) -> Result<&OsHandle> {
+    /// Return a reference to the `OsHandle` or `VirtualFile` treating it as an
+    /// actual file/dir, and allowing operations which require an actual file and
+    /// not just a stream or socket file descriptor.
+    pub(crate) fn as_file<'descriptor>(&'descriptor self) -> Result<Handle<'descriptor>> {
         match self {
-            Self::OsHandle(file) => Ok(file),
+            Self::OsHandle(handle) => Ok(Handle::OsHandle(handle)),
+            Self::VirtualFile(virt) => Ok(Handle::VirtualFile(virt.as_ref())),
             _ => Err(Error::EBADF),
         }
     }
 
     /// Like `as_file`, but return a mutable reference.
-    pub(crate) fn as_file_mut(&mut self) -> Result<&mut OsHandle> {
+    pub(crate) fn as_file_mut<'descriptor>(
+        &'descriptor mut self,
+    ) -> Result<HandleMut<'descriptor>> {
         match self {
-            Self::OsHandle(file) => Ok(file),
+            Self::OsHandle(handle) => Ok(HandleMut::OsHandle(handle)),
+            Self::VirtualFile(virt) => Ok(HandleMut::VirtualFile(virt.as_mut())),
             _ => Err(Error::EBADF),
         }
     }
 
+    pub(crate) fn as_handle<'descriptor>(&'descriptor self) -> Handle<'descriptor> {
+        match self {
+            Self::VirtualFile(virt) => Handle::VirtualFile(virt.as_ref()),
+            Self::OsHandle(handle) => Handle::OsHandle(handle),
```

Comment:
This is a workaround for another subtle bug/point of confusion: `descriptor_as_oshandle` is lossy on BSDs. BSD `OsHandle` maintains wasi-common-side state to handle `readdir` subtleties, but `descriptor_as_oshandle` constructs an `OsHandleRef` by making a new `File` from the shared fd and putting that in a new `OsHandle` it holds with `ManuallyDrop`. This discards wasi-common-side state for the file and results in incorrect `readdir` behavior.

I think generally we probably don't want `OsHandleRef` while also having state on `OsHandle`, and my first draft of this PR in fact removed the one and only place `OsHandleRef` was used, but as of last week I ran into some issues here. So I'm noting this here and if I've been convincing I'll copy the above into an issue.

Reply:
That's some great observation, thanks for pointing it out! We indeed seem to be discarding the state of `OsHandle` on BSD which contains a `DIR*` stream essentially allowing us for correctly handling seek, tell and rewinds in `fd_readdir` call. I wonder if that's a problem for `OsHandleRef` since I always assumed that would point at a stream-like file descriptor for which `fd_readdir` wouldn't really make much sense in its current form? Just thinking out loud here, it'd probably be best if @sunfishcode chipped in here as well.

Reply:
Yeah, `OsHandleRef` was motivated by the need to abstract over files and stdin/stdout/stderr, so it's aimed at stream-like use cases, and not `fd_readdir`. However, I agree that this is a really subtle issue that would be good to think through.

The new `Handle` type introduced here is a lot like `Descriptor`. Would it be feasible to just pass around `Descriptor` references, rather than create `Handle`/`HandleMut` objects to pass around? Similarly, it may make sense to pass around `Descriptor` references instead of `OsHandleRef` too, and only convert to `File` or similar at the point of calling into yanix/winx/libc.

Reply:
I've added a commit removing `Handle` and `HandleMut`: [0c571c6](https://github.com/bytecodealliance/wasmtime/pull/701/commits/0c571c6b7389ebd04d1f97ed3f64520c86f3f457). I originally added them for a similar reason as `OsHandleRef` - to abstract over the different `Stream` variants. It turns out that there really isn't much benefit of that though, either I'd call `as_os_handle()` anyway, or having any non-`File`-like is an error.

It would be nice if we could rely on the type system to enforce that anything in `sys` (or later, calling `yanix` and friends) is a non-virtual handle, a nice benefit of `OsHandle`/`OsHandleRef`. I imagine wasi-common streams will be a specific set of functionality still on top of an `OsHandle` so I think we could get away with passing around `OsHandle` internally, rather than full descriptors.